org.envirocar.app/src/org/envirocar/app/views/login/SignupActivity.java

```diff
@@ -346,4 +317,111 @@ private void showPrivacyStatementDialog() {
+        } else if (username.length() < 6) {
+            usernameEditText.setError(getString(R.string.error_invalid_username));
+            isValidUsername = false;
+        } else {
+            isValidUsername = true;
+        }
+    }
+
+    // check if the password confirm is empty
+    private void checkConfirmPassword(String password2) {
+        if (password2 == null || password2.isEmpty() || password2.equals("")) {
+            password2EditText.setError(getString(R.string.error_field_required));
+            isValidMatch = false;
+        } else {
+            isValidMatch = true;
+        }
+    }
+
+    // check if passwords match
+    private Boolean checkMatchpassword(String password, String password2) {
+        if (!password.equals(password2)) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    private void showMatchError(Boolean e) {
+        if (!e) {
+            password2EditText.setError(getString(R.string.error_passwords_not_matching));
+            password2EditText.requestFocus();
+            isValidMatch = false;
+        } else {
+            password2EditText.setError(null);
+            isValidMatch = true;
+        }
+    }
+
+    // Check for a valid email address.
+    private void checkValidEmail(String email) {
+        if (TextUtils.isEmpty(email)) {
+            emailEditText.setError(getString(R.string.error_field_required));
+            isValidEmail = false;
+        } else if (!email.matches("^[_A-Za-z0-9-\\+]+(\\.[_A-Za-z0-9-]+)*@[A-Za-z0-9-]+(\\" +
+                ".[A-Za-z0-9]+)*(\\.[A-Za-z]{2,})$")) {
+            emailEditText.setError(getString(R.string.error_invalid_email));
+            isValidEmail = false;
+        } else {
+            isValidEmail = true;
+        }
+    }
+
+    private void userNameObservable() {
+        Observable<String> observable = RxTextView.textChanges(usernameEditText).skipInitialValue().debounce(600, TimeUnit.MILLISECONDS)
+                .observeOn(AndroidSchedulers.mainThread())
+                .map(CharSequence::toString);
+        observable.subscribe(this::usernameCheck, LOG::error);
+    }
+
+    private void emailObservable() {
+        Observable<String> observable = RxTextView.textChanges(emailEditText).skip(1).debounce(600, TimeUnit.MILLISECONDS)
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribeOn(AndroidSchedulers.mainThread())
+                .map(CharSequence::toString);
+        observable.subscribe(this::checkValidEmail, LOG::error);
+    }
+
+    private void passwordObservable() {
+        Observable<String> observable = RxTextView.textChanges(password1EditText).skip(1).debounce(600, TimeUnit.MILLISECONDS)
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribeOn(AndroidSchedulers.mainThread())
+                .map(CharSequence::toString);
+        observable.subscribe(this::passwordCheck, LOG::error);
```

Comment:
don't split the observable from the subscription. Just do this
```
RxTextView.textChanges(password1EditText)
                .skipInitialValue()
                .debounce(600, TimeUnit.MILLISECONDS)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(AndroidSchedulers.mainThread())
                .map(CharSequence::toString);
                .subscribe(this::passwordCheck, LOG::error);
```
otherwise you make mistakes like the one below

Reply:
COPY the code above! and remove ```Observable<String> observable =```